Allow empty fileContent

ajhoddinott opened the pull request:
### Current behavior:

Attempting to upload an empty file throws an error.

### Desired behavior:

To be able to upload an empty file.

### Steps to reproduce

cy.get('[data-cy="file-input"]').upload({ '', fileName: 'empty.txt', mimeType: 'text/plain' });

### Versions

cypress-file-upload 3.4.0

abramenal commented:
Hi @ajhoddinott 
Well, I could not see the use case of trying to upload empty file content. Can you please describe your case?
Could it be possible solution for you:
```
const emptyFileContent = ' '; /* you just put a space as file content so that it is not nullable */

cy.get('[data-cy="file-input"]').upload({
  fileContent: emptyFileContent,
  fileName: 'empty.txt',
  mimeType: 'text/plain'
});
```

ajhoddinott commented:
Thanks for the quick response!

Cypress is for testing, and you need to be able to test edge cases:
It “doesn’t crash when the user uploads an empty file”.
It “should show an error message when the user tries to upload an empty file”.
etc.

A nearly empty file is not the same as empty file for test purposes.

abramenal commented:
@ajhoddinott seems reasonable 😄 
Then I would say we need to create a separate option which would allow that.
Something like:
```
  cy.get().upload({ fileContent, fileName, mimeType: 'application/json' }, { allowEmpty: true });
```
Can you please take a look on this snippet and update the PR accordingly?
After that I am happy to release this with new plugin version.

ajhoddinott commented:
I added the allowEmpty option as you suggested.
I also added a test for it, in a separate commit in case you want to cherry pick.
The cypress-file-upload version will need to be updated in the shadow-dom-native devDependencies to match whichever release includes allowEmpty.

abramenal commented:
@all-contributors add @ajhoddinott for code

Thanks a lot, that is perfect!

ajhoddinott commented:
Thanks for merging, and sorry for the unnecessary parameter!

One minor niggle: Your refactor from `fileContent !== undefined`  to `!fileContent` in 92382d3 changes the nuance from _may_ be empty  to _must_ be empty. I can live with it, but I preferred the _may_ interpretation, which I think is implied by the option name.

Anyway, thanks again for the quick merge and publish.

abramenal commented:
@ajhoddinott oh right, good point.
Though comparing with undefined will always give truthy value (cuz it either empty or non-empty `string`). 

So essentially it should look like 
```
const fileContentValid = allowEmpty ? true : !!fileContent;
```
or, in shorter way:
```
const fileContentValid = allowEmpty || !!fileContent;
```

ajhoddinott commented:
@abramenal Don’t you want to validate for a missing parameter?
I don’t think that has been caught earlier, so undefined is still possible. Perhaps I’m missing something.

abramenal commented:
@ajhoddinott 
If you passed `allowEmpty: true` does it make sense to check if anything passed at all?

ajhoddinott commented:
@abramenal I think so.

If a simulated user doesn’t select a file to upload, no file should be uploaded. At present, if `allowEmpty` is `true` and a test mistakenly calls for the upload anyway, the behaviour is undefined. Perhaps an empty file is uploaded. Perhaps something that relies on `fileContent` being a string throws an unexpected error. To me, it makes sense to validate this parameter just as you validate the other required parameters.

My tests now do what I want though, so I’m not strongly vested in changing it back if you disagree.

asumaran commented on recipes/shadow-dom-native/cypress/integration/file-input.spec.js:
```diff
@@ -46,4 +46,19 @@ describe('Attach file to a shadow dom input element', () => {
   });
+
+  it('successfully uploads an empty file when allowEmpty is true', () => {
+    cy.fixture('empty.txt', 'base64').then(fileContent => {
```
it looks like it's not possible to upload a *not* empty file when `allowEmpty` is `true`?

ajhoddinott replied:
See comment https://github.com/abramenal/cypress-file-upload/pull/119#issuecomment-551436057 above, and subsequent back and forth.

asumaran replied:
`allowEmpty` to me it means that an empty file can also be uploaded apart from the default behavior. Otherwise the prop should be renamed to `allowEmptyOnly` to avoid confusions.

ajhoddinott replied:
If we’re going to revisit it, I’d rather it was changed back to the original allowEmpty intent, but as I said above, I don’t feel strongly about it.
